[New Concept]: Decorators (#2838)
exercism/python: concepts/decorators/about.md

# About

Decorators are a feature of the Python language that can modify or register a function.

## How to use them

Decorators are placed just above a function in the code like so:

```py
@decorator
def function():
    pass
```

Some decorators take arguments:

> ```py

Reviewer: ```suggestion
```python

```

Reviewer: we should spell out the language name for these code blocks.

Author: I assume this is for all code blocks?

@decorator2(name="Bob")
def function2():
    pass
```

If a decorator takes any arguments, but you don't wish to pass any, you must still call it with parentheses for the decorator to work:

```py
@decorator3()
def function3():
    pass
```

Decorators are just syntactic sugar.
An alternative way of doing the same as the above three examples is using higher-order functions:
```py
def function():
    pass

function = decorator(function)

def function2():
    pass

function2 = decorator2(name="Bob")(function2)

def function3():
    pass

function3 = decorator3()(function3)
```

## How to write them

As said before, decorators can modify or register a function.
The simplest decorator however, does absolutely nothing.

Decorators are functions which return functions.
They take one argument - the function which they have been added to.
They also return one thing - the function which will be run by the program.

> This is usually either the function they were added to or a slightly modified version of that function, but it does not have to be - it can be any function whatsoever.

Reviewer: Again, I think this content is better at the top, in the first section.   Here in this section, I would suggest having a complete, simple decorator in a code example with comments to explain.  Then you can elaborate, go through the steps of writing one, and perhaps extend the original code through successive examples?   Here is an annotated starting example from the PyBit.es article:

![image](https://user-images.githubusercontent.com/5923094/148606627-b624a97b-1f11-4c86-962a-68c2925c6636.png)

The simplest decorator - one that does nothing - can be written as follows:
```py
def do_nothing(function):
    return function
```
```py
def function4a():
    return 4

@do_nothing
def function4b():
    return 4
```
```py
>>>function4a()
4
>>>function4b()
4
```
As you can see, both functions do the same thing, and when the decorator is added nothing changes.

> It is common to use the short form, `f`, instead of `function` which we will be using in the rest of the examples.

Reviewer: This _might_ be common some places.  However, in general it is bad form to use single letter variable names (even though the Python docs do so!), so we try to avoid them in our concept exercises.

A decorator that registers a function could be written as follows:
```py
functions = []

def register_function(f):
    functions.append(f)
    return f
```
```py
>>>functions
[]
```
```py
@register_function
def function5():
    return 5
```
```py
>>>functions
[<function function5 at 0x000001CA840AA700>]
>>>function5()
5

> ```

Reviewer: I like showing an example of "registering" ... but I think this could benefit from having a bit more detail.  Maybe something like what this snip from "Python Distilled" does?   Or another scenario where you're building a registry?

![image](https://user-images.githubusercontent.com/5923094/148607908-12a0514d-50be-4a20-b288-24b52c3f79b4.png)

As you can see, adding the decorator to the function adds the function into the list.

> It is important to note that this happens when the function is declared, not when it is called.

Reviewer: This is another thing that I think is important to state _up front_.  Either at the top of this "how to build" section -- or right before or after the initial examples that show decorators in use.  Here, I think it might get buried or missed by a student.

    return 6

@double
def function6b():
    return 6
```
```py
>>>function6a()
6
>>>function6b()
12
```
As you can see the decorator modifies the result of the function.

Now lets take a look at what is actually happening in this decorator.

On the first line, we have the normal function definition.
But, when we go inside this function and take a look at the next line, we get another function definition.
This function is what we will return at the end of the decorator.
This means we are not returning the original function.
Because of this, Python will not actually call the original function when a call is made to what looks like that function.
Instead, Python will call the function we created, `wrapper()`.

If we look inside our `wrapper()` function, we first see a call to the original function.
This call will do exactly the same as if we had called the original function without the decoration to the function.
Then we get a line that modifies this result.
Finally we return the modified result, which is what we get when calling the function.

If we apply this decorator to a function which takes a single number and triples it, then we should get a function that sextuples our number.

> Lets see what happens:

Reviewer: This is a good step-by-step, but it is _a lot_ to follow in one, big chunk.  I'd suggest pulling out comments like 

>This call will do exactly the same as if we had called the original function without the decoration to the function.

and using them as comments _within_ code examples, so that students see them as they are looking at the code, and then putting things like

>Finally we return the modified result, which is what we get when calling the function. 

Above or below the code example.

If function `A` is a decorator, and function `B` is its argument, then function `A` modifies, extends or replaces function `B`'s **behavior** _without modifying_ function `B`'s code.
We say that the decorator function `A` _wraps_ function `B`.
While we talk about "modifying" behavior, the wrapped function is _not actually changed_.
Behavior is either added _around_ the wrapped function (_and what it returns_), or the wrapped function's behavior is _substituted_ for some other behavior.

## A Decorator is a Higher-Order Function

A [higher-order function][higher-order functions] is a function that accepts one or more functions as arguments and/or returns one or more functions.
A function, used as an argument or returned from another function, is a [first-class function][first-class functions].
A Python function, as a [callable object][callable objects], is a first-class function which can be stored in a variable or used as an argument, much like any other value or object.
Higher-order functions and first-class functions work together to make decorators possible.

> ## What Using Decorator Looks Like

Reviewer 2: ```suggestion
## What Using a Decorator Looks Like
```

The `@` symbol is prepended to the name of the decorator function and placed just above the function to be decorated, like so:

```python
@decorator
def decorated_function():
    pass
```

Some decorators accept arguments:

```python
@decorator_with_arg(name="Bob")
def decorated_function2():
    pass
```

> If a decorator has defined default arguments, you must use parenthesis in the `@decorator()` call for the decorator to work:

Reviewer 2: Why?  We might want to explain what the problem is being presented to the parser when there is no parenthesis.  Or pose it as a "Getting to Deep Dive Python" point…

Reviewer 3: Next commit will include `, as you would in calling any function:`

The decorator then returns its inner function.
The inner function may then return the original function argument.

### A Validating Decorator Using an Inner Function

Following is an example of a decorator being used for validation:

```python
>>> def my_validator(func):
...     def my_wrapper(world):
...         print(f"Entering {func.__name__} with {world} argument")
...         if ("Pluto" == world):
...             print("Pluto is not a planet!")
...         else:
...             return func(world)
...     return my_wrapper
...
... @my_validator
... def my_func(planet):
...     print(f"Hello, {planet}!")
...
>>> my_func("World")
Entering my_func with World argument
Hello, World!
...
>>> my_func("Pluto")
Entering my_func with Pluto argument
Pluto is not a planet!
```

On the first line, we have the definition for the decorator with its `func` argument.
On the next line is the definition for the decorators _inner function_, which wraps the `func` argument.
Since the _inner function_ wraps the decorator's `func` argument, it is passed the same argument that is passed to `func`.
Note that the wrapper doesn't have to use the same name for the argument that was defined in `func`.

> The original function uses `planet` and the decorator uses `world`, and the decorater still works.

Reviewer 2: ```suggestion
The original function uses `planet` and the decorator uses `world`, and the decorator still works.
```

The inner function returns either `func` or, if `planet` equals `Pluto`, it will print that Pluto is not a planet.
It could be coded to raise a `ValueError` instead.
So, the inner function wraps `func`, and returns either `func` or does something that substitutes what `func` would do.
The decorator returns its _inner function_.

> It does not _directly_ return the original, passed-in function.

Reviewer 2: This suggests that it either must or may return the original passed in function indirectly.  This could be clarified.

Reviewer 3: The last line will be changed in the next commit to

The _inner_function_ may or may not return the original, passed-in function.

> Depending on what happens in the wrapper function, `func` may or may not be returned.

Reviewer 2: This does not, for me, clarify the prior sentence.

Reviewer 3: The whole paragraph will be rephrased in the next commit to

The inner function returns either `func` or, if `planet` equals `Pluto`, it will print that Pluto is not a planet.
It could be coded to raise a `ValueError` instead.
So, the inner function wraps `func`, and returns either `func` or does something that substitutes what `func` would do.
The decorator returns its _inner function_.
The _inner_function_ may or may not return the original, passed-in function.
Depending on what code conditionally executes in the wrapper function or _inner_function_, `func` may be returned, an error could be raised, or a value of `func`s return type could be returned.

It states and then repeats itself to hopefully be more reinforcing than obfuscatory.

### Decorating a Function that Takes an Arbitrary Number of Arguments

Decorators can be written for functions that take an arbitrary number of arguments by using the `*args` and `**kwargs` syntax.

Following is an example of a decorator for a function that takes an arbitrary number of arguments:

```python
>>> def double(func):
...     def wrapper(*args, **kwargs):
...         return func(*args, **kwargs) * 2
...     return wrapper
...
... @double
... def add(*args):
...     return sum(args)
...
>>> print(add(2, 3, 4))
18
>>> print(add(2, 3, 4, 5, 6))
40
```

This works for doubling the return value from the function argument.

> If we want to triple, quadruple, etc. the return value, we can add a parameter to the decorator itself.

Reviewer 2: ```suggestion
If we want to triple, quadruple, etc. the return value, we can add a parameter to the decorator itself., as we show in the next section.
```

Higher-order functions and first-class functions work together to make decorators possible.

## What a Decorator looks like

The `@` symbol is prepended to the name of the decorator function and placed just above the function to be decorated, like so:

```python
@decorator
def function():
    pass
```

Some decorators take arguments:

```python
@decorator_with_arg(name="Bob")
def function2():
    pass
```

If a decorator takes a default argument, you must still call it with parentheses for the decorator to work:

```python
@decorator_with_default_arg()
def function3():
    pass
```

If a decorator takes a positional argument, using empty parentheses will result in an error which will look something like:

```
TypeError: decorator_with_pos_arg() missing 1 required positional argument: 'name'
```

The `@decorator` syntax is just syntactic sugar.
Following are examples of alternative ways for setting a decorator:

```python
def function():
    pass

function = decorator(function)

def function2():
    pass

function2 = decorator_with_arg(name="Bob")(function2)

def function3():
    pass

function3 = decorator_with_default_arg()(function3)

> ```

Reviewer 3: Feel free to change it however you'd like.

## How to write a simple Decorator

Decorators are intended to extend or replace the behavior of another function, but a decorator may do nothing but return the function it is wrapping.

Decorators are functions which take at least one argument - the function which they are wrapping.

> They usually return either the wrapped function or the result of an expression that uses the wrapped function.

Reviewer 3: I don't mind them.;

The decorator returns its inner function.
The inner function may return the original function argument.

### How to write a validating Decorator using an inner function

Following is an example of a decorator being used for validation:

```python
>>> def my_validator(func):
...     def my_wrapper(planet):
...         print(f"Entering {func.__name__} with {planet} argument")
...         if ("Pluto" == planet):
...             print("Pluto is not a planet!")
...         else:
...             return func(planet)
...     return my_wrapper
... 
... @my_validator
... def my_func(planet):
...     print(f"Hello, {planet}!")
... 
>>> my_func("World")
Entering my_func with World argument
Hello, World!
>>> my_func(Pluto")
Entering my_func with Pluto argument
Pluto is not a planet!

```

On the first line, we have the definition for the decorator with its `func` argument.
On the next line is the definition for the decorators inner function, which wraps the `func` argument.
Since the inner function wraps the decorator's `func` argument, it is passed the same argument of `planet`.
It doesn't have to have the same name of `planet`.

> If `planet` was replaced with `world` throughout the decorator, the decorater would still work.

Reviewer 3: Example and text have been modified.

The inner function returns either `func` or, if `planet` equals `Pluto`, it will print that Pluto is not a planet.
It could be coded to raise a `ValueError` instead.

> So, the inner function wraps `func`, and returns either `func` or does something that substitutes what `func` would do.

Reviewer 3: I prefer my original wording.

# About

> Decorators are a feature of the Python language that can modify or register a function.

Reviewer 4: Does "modify or register" narrow the definition a bit much? There's other ways a decorator might be used which I wouldn't put into either of those buckets.

Author: I'm not really sure what I could change this to?

Author: Could this work @IsaacG?
```suggestion
Decorators are a feature of the Python language that can affect what Python actually executes when a function is called.

```

## How to use them

Decorators are placed just above a function in the code like so:

```py
@decorator
def function():
    pass
```

> Some decorator functions take arguments:

Reviewer 4: Are decorators and decorator functions two different things? This discusses decorators then decorator functions without an introduction of the term.

```py
@decorator2(name="Bob")
def function2():
    pass
```

If a decorator function takes any arguments, but you don't wish to pass any, you must still call it with parentheses for the decorator to work:

```py

> @decorator2()

Reviewer 4: Should this be `decorator3` to remain consistent? What about `decorator1` for the first?

Alternatively, just reuse the same name for the decorator and function for all examples.

Author: I went for 2 because it could well be using the same decorator function as the example above

def function3():
    pass
```

> Decorators are just syntactic sugar. An alternative way of doing the same as the above three examples is using higher-order functions:

Reviewer 4: By and large the docs split lines on sentence-ends with one sentence per line.

Author: Yeah, I realised that this one was merged just after committing, so will fix it with my next commit rather than making a single commit to fix it!

```py
def function():

>     pass

Reviewer 2: Python is definitely an indentation language, but we probably should have a blank line as well for some distinction between the function and the function called and that value assignment.

I am not sure what the style guide says (PEP8) but it looks strange to me to not have that extra blank line.

```suggestion
    pass

```

function = decorator(function)

def function2():
    pass
function2 = decorator2(name="Bob")(function2)

def function3():
    pass
function3 = decorator2()(function3)

> ```

Reviewer 2: Technical point, missing EOL (end of line) character at line 44.

Author: I'm not quite sure what you mean?

Reviewer 2: It looks like you fixed it, by appending `\r\n` to the end of the file.  But what I mean is that by definition lines have an end of line, and it is expected.  Sometimes our tools seem to forget this and because this is a text file, and text files have lines, this one character is expected at the end of each line.

# About

Decorators are a feature of the Python language that can modify or register a function.

## How to use them

Decorators are placed just above a function in the code like so:

```py
@decorator
def function():
    pass
```

Some decorators take arguments:

```py
@decorator2(name="Bob")
def function2():
    pass
```

If a decorator  takes any arguments, but you don't wish to pass any, you must still call it with parentheses for the decorator to work:

```py
@decorator3()
def function3():
    pass
```

Decorators are just syntactic sugar.
An alternative way of doing the same as the above three examples is using higher-order functions:
```py
def function():
    pass

function = decorator(function)

def function2():
    pass

function2 = decorator2(name="Bob")(function2)

def function3():
    pass

function3 = decorator3()(function3)
```

## How to write them

As said before, decorators can modify or register a function.

> The simplest decorator however, does absolutely nothing.

Reviewer 2: ```suggestion
The simplest decorator however, does absolutely nothing.
```
